### ISSUE TITLE
Got AccessDeniedException with Java 9+ on Windows 10 when trying to start CassandraUnit

### ISSUE THREAD

**Issue by gitaroktato** [jsevellec/cassandra-unit#283]
With any JDK after Java 8, CassandraUnit fails to start with AccessDeniedException on Windows 10.

After spending some investigation we suspect that `WindowsNativeDispatcher.DeleteFile0` implementation changed and behaving differently. Another thread is holding the file handle for the commit log CassandraUnit tries to delete. Expected behavior should be to delete the file after the handle is released (see this StackOverflow answer for details: https://stackoverflow.com/a/31608180/8321787)

This is the repo we used for testing: https://github.com/peterkarsa/working-cassunit-ubuntu (thanks to @peterkarsa) for putting it together and helping with troubleshooting)

```
FSWriteError in target\embeddedCassandra\commitlog\CommitLog-6-1542961092250.log
	at org.apache.cassandra.io.util.FileUtils.deleteWithConfirm(FileUtils.java:142)
	at org.apache.cassandra.io.util.FileUtils.deleteWithConfirm(FileUtils.java:159)
	at org.apache.cassandra.db.commitlog.CommitLogSegment.discard(CommitLogSegment.java:409)
	at org.apache.cassandra.db.commitlog.AbstractCommitLogSegmentManager.discardAvailableSegment(AbstractCommitLogSegmentManager.java:473)
	at org.apache.cassandra.db.commitlog.AbstractCommitLogSegmentManager.shutdown(AbstractCommitLogSegmentManager.java:460)
	at org.apache.cassandra.db.commitlog.AbstractCommitLogSegmentManager.stopUnsafe(AbstractCommitLogSegmentManager.java:405)
	at org.apache.cassandra.db.commitlog.CommitLog.stopUnsafe(CommitLog.java:445)
	at org.apache.cassandra.db.commitlog.CommitLog.resetUnsafe(CommitLog.java:418)
	at org.cassandraunit.utils.EmbeddedCassandraServerHelper.cleanupAndLeaveDirs(EmbeddedCassandraServerHelper.java:329)
	at org.cassandraunit.utils.EmbeddedCassandraServerHelper.startEmbeddedCassandra(EmbeddedCassandraServerHelper.java:145)
	at org.cassandraunit.utils.EmbeddedCassandraServerHelper.startEmbeddedCassandra(EmbeddedCassandraServerHelper.java:108)
	at org.cassandraunit.utils.EmbeddedCassandraServerHelper.startEmbeddedCassandra(EmbeddedCassandraServerHelper.java:88)
	at org.cassandraunit.utils.EmbeddedCassandraServerHelper.startEmbeddedCassandra(EmbeddedCassandraServerHelper.java:84)
	at com.example.demo.DemoApplicationTests.contextLoads(DemoApplicationTests.java:20)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:583)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:719)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:989)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at org.testng.TestRunner.privateRun(TestRunner.java:648)
	at org.testng.TestRunner.run(TestRunner.java:505)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:455)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:450)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:415)
	at org.testng.SuiteRunner.run(SuiteRunner.java:364)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:84)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1208)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1137)
	at org.testng.TestNG.runSuites(TestNG.java:1049)
	at org.testng.TestNG.run(TestNG.java:1017)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:123)
Caused by: java.nio.file.AccessDeniedException: target\embeddedCassandra\commitlog\CommitLog-6-1542961092250.log

	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:89)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:270)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:105)
	at java.base/java.nio.file.Files.delete(Files.java:1141)
	at org.apache.cassandra.io.util.FileUtils.deleteWithConfirm(FileUtils.java:136)
	... 37 more
```

**Comment by prashantbhardwaj** [comment]
We are using CassandraCQLUnit in 2 separate test classes. Individual files have many tests. All of these tests are passing when we run each file independently. However when tests of these 2 files get executed during maven clean install, tests of one class pass and other class fail with same error. AccessDenied on commitLog file. Any help please?

**Comment by jarpz** [comment]
Hi @jsevellec 

I'm getting the same error when I compile my project with java 9+ on Windows machines. It's very rare because with java 8 it works perfectly.. Do you have any clue for solve this problem?

**Comment by finrod2002** [comment]
I use cassandraUnit for my integration tests, which I run both on Windows, Linux and also on Windows Linux subsystem. Only in Windows I see the AccessDeniedException when using JDK 11. Linux and Windows Linux subsystem (which I guess can be considered equivalent environments in this context) are working fine.

The exception in Windows is the same as reported by the OP.

For my use cases where I don't have to delete entire tables and keyspaces during test runs. I was successful by applying the following quick and dirty workaround.

1) First I check if the tests run on a Windows system, if not I just start everything "normally" by using EmbeddedCassandraServerHelper. 
2) If tests are run on Windows instead of starting cassandra with EmbeddedCassandraServerHelper I use my custom WindowsEmbeddedCassandraServerHelper class with the only modification in the following method.

` private static void cleanupAndLeaveDirs() throws IOException
    {
        mkdirs();
        cleanup();
        mkdirs();
        final CommitLog commitLog = CommitLog.instance;

        // this is the custom Windows workaround: we catch FSWriteError
        try
        {
            commitLog.stopUnsafe(true);
        }
        catch (final FSWriteError ae)
        {
        }
        commitLog.resetConfiguration();
        commitLog.restartUnsafe();
    }`

So basically I just catch the FSWriteError. I know it's quick and dirty and not nice, but for my use case good enough and I hope that it can help somebody else.

**Comment by DBarbarian** [comment]
I don't know if this helps in any way, but we have recently upgraded to CassandraUnit 4.3.1.0 and the problem still exists there. Also I have reproduced the issue in Linux when running on a mounted Windows disk.

**Comment by amitdivyansh** [comment]
Did this issue resolved? Can we have update on this please?

**Comment by sampathkdhandu** [comment]
Any update on this issue?

**Comment by jsevellec** [comment]
**I could not reproduce this, so treat what follows as a likely fix rather than a verified one.**

`AccessDeniedException` on Windows with Java 9+. Two changes are relevant: the config path is now turned into a URI with `file.toPath().toUri()` instead of hand-built `file://` / `file:/` string concatenation (which was wrong for drive letters), and `tmpDir` now actually relocates the storage directories instead of leaving them under a CWD-relative `target/`.

Closing on that basis to get the tracker honest, but please **reopen** if it still happens against 5.0.0 — a reopened issue with a stack trace is much more useful than this one sitting open and unanswered for years. 5.0.0 is on `main` but not yet on Maven Central.

Sorry for the long silence.
